### PR TITLE
ath79: add support for NEC Aterm WG800HP

### PR DESCRIPTION
**Pull request by musashino205**
This PR includes following changes:

- fix pinmux reg value for QCA956x SoC
  - qca9557.dtsi has same issue?
- add "hex pattern mode" for xorimage
- add support for NEC Aterm WG800HP

### CODE REVIEW

File: target/linux/brcm63xx/image/Makefile [hunk 1]
```diff
@@ -132,11 +132,6 @@ define Build/spw303v-bin
 	mv $@.spw303v $@
 endef
 
-define Build/xor-image
```
Comment (ynezz) [inline target/linux/brcm63xx/image/Makefile]: @musashino205 This is good idea in itself, but hidden in the patch for new board in `ath79` platform, althought touching code in `brcm63xx` platform. So it probably belongs to separate patch for reusable `xor-image` image target.

File: tools/firmware-utils/src/xorimage.c [hunk 2]
```diff
@@ -24,7 +24,11 @@
 #include <unistd.h>
 #include <sys/stat.h>
 
+#define TRUE 1
```
Comment (chunkeey) [inline tools/firmware-utils/src/xorimage.c]: Why not C99's [stdbool.h](https://en.wikipedia.org/wiki/C_data_types#stdbool.h)?

Comment (musashino205) [reply tools/firmware-utils/src/xorimage.c]: I do not have much knowledge about C language.
There are two methods of adding define for TRUE/FALSE like *srec2bin.c* and including ```stdbool.h``` like *mktplinkfw.c*, so I did not know which to use.

Should I use stdbool.h?

Comment (chunkeey) [reply tools/firmware-utils/src/xorimage.c]: > Should I use stdbool.h?

I think so, since you are using the "for' loop initial declarations" (a C99 feature) a bit further down in line 118 of the same file.

File: target/linux/ath79/dts/qca9563_nec_wg800hp.dts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+	leds {
+		compatible = "gpio-leds";
+		pinctrl-names = "default";
+		pinctrl-0 = <&jtag_disable_pins>;
+
+		wlan2g_red {
+			label = "wg800hp:red:wlan2g";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
Comment (chunkeey) [inline target/linux/ath79/dts/qca9563_nec_wg800hp.dts]: The "off" state is already the default. You don't need to explicitly set it.

Though, I'm fine with it. Problem is: Somebody is bound to go on a crusade and remove all "default-state = off".

File: tools/firmware-utils/src/xorimage.c (hunk 2 above)
Comment (musashino205) [reply tools/firmware-utils/src/xorimage.c]: OK, I'll modify it.

File: target/linux/ath79/dts/qca9563_nec_wg800hp.dts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+		phy-mode = "sgmii";
```
[the remark this answers is not in the document]
Comment (musashino205) [reply target/linux/ath79/dts/qca9563_nec_wg800hp.dts]: I see. I'll drop it.

File: target/linux/ath79/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 5]
```diff
@@ -96,6 +96,10 @@ case "$FIRMWARE" in
 		ath10kcal_extract "art" 20480 2116
 		ath10kcal_patch_mac $(macaddr_add $(mtd_get_mac_ascii u-boot-env ethaddr) +1)
 		;;
+	nec,wg800hp)
+		ath10kcal_extract "art" 20480 2116
+		ath10kcal_patch_mac $(mtd_get_mac_text board_data 2176)
```
[the remark this answers is not in the document]
Comment (musashino205) [reply target/linux/ath79/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: OK, I'll try it.

File: target/linux/ath79/dts/qca9563_nec_wg800hp.dts (hunk 3 above)
Comment (musashino205) [reply target/linux/ath79/dts/qca9563_nec_wg800hp.dts]: Oh, I see. I'll drop it.

**Comment by musashino205** [comment]
Done.

File: target/linux/ath79/dts/qca9563_nec_wg800hp.dts (hunk 3 above)
Comment (chunkeey) [reply target/linux/ath79/dts/qca9563_nec_wg800hp.dts]: > Problem is: Somebody is bound to go on a crusade and remove all "default-state = off".

And It just happened. 

https://patchwork.ozlabs.org/patch/1014114/

Good timing.

File: tools/firmware-utils/src/xorimage.c [hunk 6]
```diff
@@ -100,6 +107,17 @@ int main(int argc, char **argv)
...
 
+	if (is_hex_pattern) {
+		if (p_len % 2 != 0) {
+			fprintf(stderr, "the number of characters (hex) is incorrect\n");
+			usage();
+		}
+
+		for (int i = 0; i < (p_len / 2); i++) {
+			sscanf(pattern + (i * 2), "%2x", &hex_buf);
+			hex_pattern[i] = (char)hex_buf;
```
Comment (chunkeey) [inline tools/firmware-utils/src/xorimage.c]: Because hex_pattern is declared as

      char *hex_pattern = default_pattern; // = "12345678";

the hex_pattern can store at most 8 bytes.

So, this can/will cause a "segmentation fault", if the provided hex pattern is big enough.

    $ ./xorimage -i xorimage -o xorimage-xor -x -p "$(seq 0 8000)"
    Segmentation fault

File: target/linux/ath79/dts/qca9563_nec_wg800hp.dts (hunk 3 above)
Comment (musashino205) [reply target/linux/ath79/dts/qca9563_nec_wg800hp.dts]: Oh...

File: tools/firmware-utils/src/xorimage.c (hunk 6 above)
Comment (musashino205) [reply tools/firmware-utils/src/xorimage.c]: Hmm...sorry, I confused...

As a pure question (due to ignorance), why is it no problem to pass a 32-byte string to ```pattern``` declared similarly by an 8-byte character string?
In that case, when converted from a character string to a hexadecimal value and passed to ```hex_pattern```, eventually it has a 16-byte value...

Comment (musashino205) [reply tools/firmware-utils/src/xorimage.c]: Thanks for your detailed advice, it seems I got a rough understanding...
Then, how long should I limit ```hex_pattern``` to?

Comment (chunkeey) [reply tools/firmware-utils/src/xorimage.c]: > Then, how long should I limit hex_pattern to?

There's already kind of a limit. Since the pattern is passed as a parameter to the xorimage program, the pattern has to go through a execve system call which enforces a string length
check of MAX_ARG_STRLEN.

https://elixir.bootlin.com/linux/v4.20-rc7/source/include/uapi/linux/binfmts.h#L15
    /*
     * These are the maximum length and maximum number of strings passed to the
     * execve() system call.  MAX_ARG_STRLEN is essentially random but serves to
     * prevent the kernel from being unduly impacted by misaddressed pointers.
     * MAX_ARG_STRINGS is chosen to fit in a signed 32-bit integer.
     */
    #define MAX_ARG_STRLEN (PAGE_SIZE * 32)

On a 4K PAGE_SIZE System this would be 131072 / 128 KIB.

Comment (musashino205) [reply tools/firmware-utils/src/xorimage.c]: Hmm...sorry, I searched and investigated about ```execve```, but I did not know how to use ```MAX_ARG_STRLEN``` through it from C source...

Comment (chunkeey) [reply tools/firmware-utils/src/xorimage.c]: > Hmm...sorry, I searched and investigated about `execve`, but I did not know how to use `MAX_ARG_STRLEN` through it from C source...

Hm, I'm getting the feeling that we are unfortunately talking past each other at this point :sob: .

The pattern string can't exceed MAX_ARG_STRLEN by design. So you could just allocate a memory buffer on the heap which holds (strlen(pattern) / 2) bytes and free it after use. But you could also declare something like char hex_pattern[128] buffer and check that (strlen(pattern) / 2) does not exceed the sizeof(hex_pattern). Or you can come up with another option. As long as it passes the fuzzing I do it's "just fine" :+1: .

Comment (musashino205) [reply tools/firmware-utils/src/xorimage.c]: I see...
Then, how about with the following changes? (Almost as your advice is...)
```char *hex_pattern = default_pattern;``` -> ```char hex_pattern[128];```

and add
```
		if ((p_len / 2) >= sizeof(hex_pattern)) {
			fprintf(stderr, "provided hex pattern is too long\n");
			usage();
		}
```

**Comment by musashino205** [comment]
Fixed xorimage.c and rebased from master.

**Comment by chunkeey** [comment]
Merged. Thanks.
